libs/langchain/langchain/indexes/_api.py

```diff
@@ -329,14 +331,20 @@ def index(
                 num_deleted += len(uids_to_delete)
 
     if cleanup == "full":
-        uids_to_delete = record_manager.list_keys(before=index_start_dt)
-
-        if uids_to_delete:
-            # First delete from record store.
-            vector_store.delete(uids_to_delete)
-            # Then delete from record manager.
-            record_manager.delete_keys(uids_to_delete)
-            num_deleted = len(uids_to_delete)
+        while True:
+            uids_to_delete = record_manager.list_keys(
+                before=index_start_dt, limit=cleanup_batch_size
+            )
+
+            if not uids_to_delete:
+                break
+
+            if uids_to_delete:
+                # First delete from record store.
+                vector_store.delete(uids_to_delete)
+                # Then delete from record manager.
+                record_manager.delete_keys(uids_to_delete)
+                num_deleted += len(uids_to_delete)
```

Comment:
nit / up to you
```suggestion
        uids_to_delete = record_manager.list_keys(
            before=index_start_dt, limit=cleanup_batch_size
        )
        while uids_to_delete:
              # First delete from record store.
              vector_store.delete(uids_to_delete)
              # Then delete from record manager.
              record_manager.delete_keys(uids_to_delete)
              num_deleted += len(uids_to_delete)
              
              uids_to_delete = record_manager.list_keys(
                  before=index_start_dt, limit=cleanup_batch_size
              )

```

Reply:
could even:
```python
while uids_to_delete := record_manager...
```

Reply:
Isn't the `while True` more explicit about the fact that there is a potential for an infinite loop? I don't mind either way though